ReadOnlyEngineTests testSearcherId fails

**Build scan**:

https://gradle-enterprise.elastic.co/s/j4tztllphmnpo

**Repro line**:
./gradlew ':server:test' --tests "org.elasticsearch.index.engine.ReadOnlyEngineTests.testSearcherId" \
  -Dtests.seed=5F23C74BF9BA02D2 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-AR \
  -Dtests.timezone=America/Noronha \
  -Druntime.java=11

**Reproduces locally?**:
yes on 7.x;   no failure on master (not in CI either)

**Applicable branches**:
only 7.x

**Failure history**:
14 failures over the last week

**Failure excerpt**:
```
org.elasticsearch.index.engine.ReadOnlyEngineTests > testSearcherId FAILED
    java.lang.AssertionError:
    Expected: not "0b4176a0054fce403abb72350d27706d167ec061d30031a49e5db37e4a50113b"
         but: was "0b4176a0054fce403abb72350d27706d167ec061d30031a49e5db37e4a50113b"
        at __randomizedtesting.SeedInfo.seed([5F23C74BF9BA02D2:BDA0039B9DAD7920]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.index.engine.ReadOnlyEngineTests.testSearcherId(ReadOnlyEngineTests.java:373)
```

Another one for 7.x: https://gradle-enterprise.elastic.co/s/5iuefv3btss72

```
java.lang.AssertionError: |  
-- | --
  | Expected: not "45be586f712ec20585769456283fa5c4c17cf1d080579342d34c266754a2c10d" |  
  | but: was "45be586f712ec20585769456283fa5c4c17cf1d080579342d34c266754a2c10d"
```

Reproducible locally with:
> ./gradlew ':server:test' --tests "org.elasticsearch.index.engine.ReadOnlyEngineTests.testSearcherId" -Dtests.seed=70ADB7A5B6AE145E -Dtests.security.manager=true -Dtests.locale=id-ID -Dtests.timezone=UCT -Druntime.java=8

The failure is seed specific, so I am *not* muting it.

--- a/server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java
+++ b/server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java
@@ -22,12 +22,10 @@ import org.apache.lucene.index.CodecReader;
 import org.apache.lucene.index.DirectoryReader;
 import org.apache.lucene.index.FilterMergePolicy;
 import org.apache.lucene.index.IndexReader;
-import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.index.MergePolicy;
 import org.apache.lucene.index.NoMergePolicy;
 import org.apache.lucene.util.IOSupplier;
 import org.apache.lucene.util.LuceneTestCase;
-import org.apache.lucene.util.SetOnce;
 import org.elasticsearch.action.admin.indices.forcemerge.ForceMergeRequest;
 import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.bytes.BytesArray;
@@ -40,7 +38,9 @@ import org.elasticsearch.index.store.Store;
 import org.elasticsearch.index.translog.TranslogStats;
 
 import java.io.IOException;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 
@@ -345,14 +345,9 @@ public class ReadOnlyEngineTests extends EngineTestCase {
                 }
             };
             final EngineConfig config = config(defaultSettings, store, createTempDir(), mergePolicy, null, null, globalCheckpoint::get);
-            final SetOnce<IndexWriter> indexWriter = new SetOnce<>();
-            final IndexWriterFactory indexWriterFactory = (dir, iwc) -> {
-                final IndexWriter writer = new IndexWriter(dir, iwc);
-                indexWriter.set(writer);
-                return writer;
-            };
             String lastSearcherId;
-            try (InternalEngine engine = createEngine(indexWriterFactory, null, null, config)) {
+            Map<String, Engine.Operation> latestOps = new HashMap<>();
+            try (InternalEngine engine = createEngine(null, null, null, config)) {
                 lastSearcherId = ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos());
                 assertNotNull(lastSearcherId);
                 int iterations = randomIntBetween(0, 10);
@@ -360,22 +355,44 @@ public class ReadOnlyEngineTests extends EngineTestCase {
                     assertThat(ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos()), equalTo(lastSearcherId));
                     final List<Engine.Operation> operations = generateHistoryOnReplica(between(1, 100),
                         engine.getProcessedLocalCheckpoint() + 1L, false, randomBoolean(), randomBoolean());
-                    final long indexWriterSeqNo = indexWriter.get().getMaxCompletedSequenceNumber();
-                    applyOperations(engine, operations);
-                    if (indexWriter.get().getMaxCompletedSequenceNumber() == indexWriterSeqNo) {
-                        assertFalse(
-                            "we don't add documents for no-ops or deletes of non existing documents iff soft-deletes is disabled",
-                            config.getIndexSettings().isSoftDeleteEnabled());
-                        continue;
+                    boolean hasChanges = false;
+                    for (Engine.Operation op : operations) {
+                        if (op instanceof Engine.Index) {
+                            engine.index((Engine.Index) op);
+                            if (latestOps.containsKey(op.id()) == false || latestOps.get(op.id()).seqNo() < op.seqNo()) {
+                                latestOps.put(op.id(), op);
+                                hasChanges = true;
+                            }
+                        } else if (op instanceof Engine.Delete) {
+                            engine.delete((Engine.Delete) op);
+                            final Engine.Operation lastOp = latestOps.get(op.id());
+                            hasChanges |= (lastOp instanceof Engine.Index && lastOp.seqNo() < op.seqNo());
+                            if (lastOp == null || lastOp.seqNo() < op.seqNo()) {
+                                latestOps.put(op.id(), op);
+                            }
+                        } else {
+                            assertThat(op, instanceOf(Engine.NoOp.class));
+                            engine.noOp((Engine.NoOp) op);
+                        }
+                        if (randomInt(100) < 10) {
+                            engine.refresh("test");
+                        }
+                        if (rarely()) {
+                            engine.flush(false, true);
+                        }
                     }
                     engine.flush(randomBoolean(), true);
-                    final String newCommitId = ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos());
-                    assertThat(newCommitId, not(equalTo(lastSearcherId)));
+                    final String newSearcherId = ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos());
+                    if (hasChanges || config.getIndexSettings().isSoftDeleteEnabled()) {
+                        assertThat(newSearcherId, not(equalTo(lastSearcherId)));
+                    } else {
+                        assertThat(newSearcherId, equalTo(lastSearcherId));
+                    }
                     if (randomBoolean()) {
                         engine.flush(true, true);
-                        assertThat(ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos()), equalTo(newCommitId));
+                        assertThat(ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos()), equalTo(newSearcherId));
                     }
-                    lastSearcherId = newCommitId;
+                    lastSearcherId = newSearcherId;
                 }
                 globalCheckpoint.set(engine.getProcessedLocalCheckpoint());
             }

